### ISSUE TITLE
Support same-state transitions out of the box via new schedules

### ISSUE THREAD

**Issue by benfrankel** [bevyengine/bevy#15478]
## What problem does this solve or what need does it fill?

Support same-state transitions out of the box for things like "restart game" or "reset UI". See https://github.com/bevyengine/bevy/issues/14447 for more information.

## What solution would you like?

Provide the following schedules:

- `OnExit(S)`: Runs on the exit side of a transition `S -> _`.
- `OnEnter(S)`: Runs on the enter side of a transition `_ -> S`.
- `OnTransition(S, T)`: Runs between the exit and enter sides of a transition `S -> T`.
- `OnReExit(S)`: Runs on the exit side of a transition `S -> S`.
- `OnReEnter(S)`: Runs on the enter side of a transition `S -> S`
- `OnChangeExit(S)`: Runs on the exit side of a transition `S -> !S`
- `OnChangeEnter(S)`: Runs on the enter side of a transition `!S -> S`

Schedule names are bikesheddable.

## What alternative(s) have you considered?

Notice that each schedule actually encodes 2 separate concepts:

1. **Ordering (when the systems run):** Do the systems run on the exit side, enter side, or in between? This is necessary so that teardown systems run before setup systems.
2. **Condition (whether the systems run):** Which state transitions cause the systems to run? Is it a transition that exits a specific state `S`? Or a transition from `S` to `S`? Etc.

This can be split into two orthogonal axes using system sets and run conditions respectively. This allows you to schedule a system to run on the enter side of a `S -> _` transition for example, or even use state pattern-matching to match on any class of transitions you're interested in.

This approach has a proof of concept in `pyri_state` that maintains comparable ergonomics, but it would require a bigger internal / breaking change to implement in `bevy_state`.

**Comment by MiniaczQ** [comment]
I'd rather keep the name of `OnExit` (here `OnChangeExit`) and `OnEnter` (here `OnChangeEnter`) and instead use `OnAnyExit`(here `OnExit`) and `OnAnyEnter(here `OnEnter`).
It's prevent the breaking change and I still think the current `OnExit` is the best default behavior, so the name should be shorter.

**Comment by shanecelis** [comment]
I'm against putting this in main. I also haven't seen a compelling motivation for why one would want to handle identity transitions. Show me the states and conditions where this is warranted. It takes a no-op in the current implementation and makes it an "op".

It adds to the API surface in a way that is going to be difficult to explain in the documentation why you'd choose one schedule over another without having to explain "state transitions" also means "state identity transitions", which aren't really transitions.

If people truly want to adopt that complexity, let them. It can be implemented by a third-party crate and has an [implementation](https://github.com/MiniaczQ/bevy/blob/main/examples/state/custom_transitions.rs) in the examples showing the way.

**Comment by benfrankel** [comment]
I could provide many examples where this is desirable. The most obvious example is restarting a game or level, or refreshing some UI that sets itself up upon entering a state. Another example is selecting a level to enter in the pause menu of a puzzle game, which may happen to be the level you're already in. "Exit the current state and then enter it again" is very natural, and not supporting it creates a weird edge case.

**Comment by shanecelis** [comment]
Thank you for providing examples. The level selection is a good example. Let's work with that. Assuming you're using a state for the level, changing it to the same level currently is a no-op, which is what motivates an identity state transition. But I believe there's another verb here, which is you also want to reset the level, which can happen to and in any level at any time, which is not a state; it's an event.

Now you may object that resetting a level is tearing it down and recreating it. But for some games, like sokoban for instance, resetting a level can just be resetting a few transforms to their original positions. Resetting can be much cheaper than tearing down and setting back up. Here's my pseudo-code to facilitate the discussion.

```rust
#[derive(States)]
pub enum GameState {
    Level(pub u8),
}
// ...
        .add_systems(OnExit(GameState::Level), teardown_level)
        .add_systems(OnEnter(GameState::Level), setup_level)

// Let's say there are two ways to deal with level resets.
// 1. Brute force approach. (Expensive)
        .add_systems(Update, (teardown_level, setup_level).chain().run_if(on_event(ResetLevel)))

// 2. Minimal dynamic state. (Cheap)
        .add_systems(Update, reset_player_and_elements.run_if(on_event(ResetLevel)))
```

For me this reinforces my point. A state machine is a good fundamental building block, but it is in its own way opinionated about what constitutes a state: You can only be in one state, and you can only transition to another state. Often times trying to use a state machine forces you to recon with latent or implicit states; this is good! It helps your design become more explicit.

Perhaps we should consider creating an example that shows this kind of level selection pattern using states and a reset event. I notice there aren't any examples with states that have additional data to them like `Level(pub u8)`.

**Comment by benfrankel** [comment]
I don't consider performance to be a legitimate issue in the level selector example. If the "level 1 -> 2" transition logic is considered sufficiently performant, then reusing that logic for "level 1 -> 1" would be as well. If somehow that's not the case, this proposal lets you micro-optimize by isolating same-state transitions anyways.

A same-state transition is still a state transition, it's not a totally different concept. Mathematically it's a self-edge. In practice it means exiting the current state and then entering the current state instead of a different state. It's natural and convenient, and treating it like a special case is an unnecessary papercut.

**Comment by shanecelis** [comment]
Well, I've been subjected to enough single-player games where dying was penalized mainly by waiting for a game to reload its level that I consider it a valid performance issue.

**Comment by benfrankel** [comment]
Sure, you'd want to restrict a system that loads the next level to only run when it differs from the previous level. That would go in `OnChangeEnter(S)`, which is why that behavior needs to still be possible. Personally I'd argue for even _more_ flexibility via run conditions instead of schedules, because I absolutely agree that these tools are necessary.

**Comment by shanecelis** [comment]
I relent that self-edges are permissible and prevalent in finite state machines. But they don't do anything mathematically. We have to choose leave them as no-ops or make them do something.

Well, let's look at your proposal given the level loading example and the cheap and expensive level reset.

```rust
// Let's say there are two ways to deal with level resets.
// 1. Brute force approach. (Expensive)
    .add_systems(OnExit(GameState::Level), teardown_level)
    .add_systems(OnEnter(GameState::Level), setup_level)

// 2. Minimal dynamic state. (Cheap)
    .add_systems(OnChangeExit(GameState::Level), teardown_level)
    .add_systems(OnChangeEnter(GameState::Level), setup_level)
    .add_systems(OnEnter(GameState::Level), reset_player_and_elements)
```

Now, I don't mind the look of either, but in the second cheap way (preferred), there is a issue we have to contend with. What order do the `Enter`s run? `Any`, `Change`, `Re`. Can they be ordered in the general case? I fear they can't be. Here `reset_player_and_elements` must come after `setup_level` so we can by fiat declare that the ordering will be `Change`, `Any`, `Re`, but will that work for everyone, all the time?

**Comment by benfrankel** [comment]
Yeah, regarding ordering you're right, and a user will have the same issue if they follow the `custom_transitions` example themselves. This is a good reason for the engine to provide a proper solution out of the box (not necessarily this particular proposal I guess, I'd just hoped it would be more palatable because it already exists in an upstream example).

The system set / run condition approach doesn't have the ordering issue OTOH because putting a run condition on a system doesn't force it into a different system set, whereas here the two concepts are awkwardly merged along a single axis.

**Comment by CalinZBaenen** [comment]
> Show me the states and conditions where this is warranted. It takes a no-op in the current implementation and makes it an "op".

I want to refresh my game's menu – which I load by listening for an `OnEnter(Screen::Menu( Menu::ThisMenusName ))` state transition – whenever the language changes.  
One may additionally want to refresh the menu for other reasons.

Similar refreshing logic could be used in games for other things that may not necessarily be loading-related.

I'm sure others here may have other, better reasons for wanting to capture the "identity state transition" (whatever the fuck that means).  
&nbsp;(I presume it just means the transition from one state to itself, kind ov like how the number one in math is the multiplicative identity.)

**Comment by shanecelis** [comment]
You can have it both ways. You can listen for state changes and you can exercise the same system when there hasn't been any state changes.

```rust
app.add_systems(Update, refresh_mesh.run_if(on_event::<RefreshMenuEvent>()));
```

* * *

Actually, this is a moot point. Bevy 0.16 supports identity state transitions with [StateTransitionEvent](https://docs.rs/bevy/latest/bevy/prelude/struct.StateTransitionEvent.html) and `OnTransition`. However, it ignores it for `OnEnter` and `OnExit`. Not a bad way to deal with it.

**Comment by shanecelis** [comment]
In light of same-state transitions being supported, I'd suggest closing this issue.

**Comment by CalinZBaenen** [comment]
> You can have it both ways. You can listen for state changes and you can exercise the same system when there hasn't been any state changes.
> 
> ```rust
> app.add_systems(Update, refresh_mesh.run_if(on_event::<RefreshMenuEvent>()));
> ```

Right, but doesn't this have the overhead ov running a check every frame?  
Also, I've never used events, and I'm not very experienced on how to use them – and, while I know I will probably have to interact with them (or know how to use them) at some point, it's quite frankly simpler to interact with states by resetting the current menu state, rather than introducing an entirely new concept to the project.

Also, using `OnTransition`, which I'm guessing one could be use to detect the change from one state to another is much simpler in premise than setting this up (in addition to the actual code that registers the events).

> In light of same-state transitions being supported, I'd suggest closing this issue.

You say it's supported... in Bevy 0.16, but I don't think Bevy 0.16 has been released, so this would be incorrect.

**Comment by shanecelis** [comment]
I misspoke. It's in Bevy 0.15.

**Comment by CalinZBaenen** [comment]
How come it's not `Option<S>` instead ov `S` (for the field values ov `OnTransition`)?  
What if I want to listen for the transition from any state to `entered`?

You might say "Then use `OnEnter`.";  however, I can not use `OnEnter` since it (still) does not handle identity transitions, I believe.

**Comment by shanecelis** [comment]
You can have them if you want them, but I still think it's not worth doing because you're taking a state machine and asking it to treat some state, say, `Menu` as both `MenuFresh` and `MenuDirty` when what's preferable is to let the state machine help you. If something makes your menu dirty such that it requires a refresh then that should something you capture in your states. Then you can write your code that deals with that explicit state instead of transitioning to the same state you're already in and doing a bunch of work that is sometimes required (when it's dirty) and sometimes not (when it's not dirty). 

Every same-state transition that requires an "operation" is a latent, unexplicated state waiting to be exposed. Expose it. 

```rust
/// ```ignore
/// app.add_system(Update, refresh_mesh.run_if(on_any_enter(AppState::Menu)));
/// ```
pub fn on_any_enter<S: States>(state: S) -> impl Fn(EventReader<StateTransitionEvent<S>>) -> bool {
    move |mut reader: EventReader<StateTransitionEvent<S>>| {
        reader.read().any(|e| e.entered.as_ref().map(|enter| *enter == state).unwrap_or(false))
    }
}

/// ```ignore
/// app.add_system(Update, teardown_mesh.run_if(on_any_exit(AppState::Menu)));
/// ```
pub fn on_any_exit<S: States>(state: S) -> impl Fn(EventReader<StateTransitionEvent<S>>) -> bool {
    move |mut reader: EventReader<StateTransitionEvent<S>>| {
        reader.read().any(|e| e.exited.as_ref().map(|exit| *exit == state).unwrap_or(false))
    }
}
```

**Comment by CalinZBaenen** [comment]
> If something makes your menu dirty such that it requires a refresh then that's something you should capture in your states.

But nothing is making my menu "dirty".

No new UI elements are added, nor are any removed – the menu is simply outdated and needs a simple refresh, which I do not believe is "a big enough deal" to need to capture with states.

> Then you can write your code that deals with that explicit state instead of transitioning to the same state you're already in and doing a bunch of work that is sometimes required (when it's dirty) and sometimes not (when it's not dirty).

Yes, but then I have to write boilerplate "bounceback" code, to to **literally nothing** ... **except** change the state back to one referring it.  
(And that's **even more** bothersome if I am using this designpattern for multiple types which implement `States`)

Ideally, Bevy would come prepacked with a way to handle this situation.  

\[To @cart: Take notes, my man.&nbsp;&nbsp;😉)

**Comment by shanecelis** [comment]
> No new UI elements are added, nor are any removed – the menu is simply outdated and needs a simple refresh, which I do not believe is "a big enough deal" to need to capture with states.

Can you share an example of what a "refresh" of your menu does? What changes?

**Comment by CalinZBaenen** [comment]
> Can you share an example of what a "refresh" of your menu does?

It reconstructs the entire menu so that any influential values (such as the values ov particular `Res`ources) may affect the menu.

See the images in [my post on Bluesky](https://bsky.app/profile/kattytheenby.bsky.social/post/3lfmeyrz4s223):
![A screenshot of the title-screen in English.](https://cdn.bsky.app/img/feed_fullsize/plain/did:plc:m2cat6yxr4v3csw2dtlxm2kt/bafkreidaw5vpituvbrjdslbrekwo337g2ysqamoppbjnyihumw7gcohvtm)
![A screenshot of the title-screen in Swedish.](https://cdn.bsky.app/img/feed_fullsize/plain/did:plc:m2cat6yxr4v3csw2dtlxm2kt/bafkreigf4dhrrcjiep3g6cuqbesfvlxbxupygeazvokzla5riyswcpsz7i)

> What changes?

**In this particular case**, the language changes, causing textual elements to update their text — however, in theory, anything could change based on what the value ov certain `Res`ources or `State`s are, since a system request specific information is responsible for setting it up.

**Comment by shanecelis** [comment]
Thank you. That's helpful. And that's cool that you're supporting multiple languages. Perhaps I'm being too stackoverflow-y about state transitions. You should be able to do it how you want to do it.

If you want to do it via same-state transitions, you can do it this way with the above code.

```rust
// OLD
// app.add_systems(OnEnter(AppState::Menu), update_menu);
// NEW
app.add_systems(Update, update_menu.run_if(on_any_enter(AppState::Menu));
```

**Comment by CalinZBaenen** [comment]
> Perhaps I'm being too stackoverflow-y about state transitions.

I'm (also) a bit stubborn about code, particularly in regards to how reusable it is.

Hopefully, though, this situation has illuminated the scenarios in which having something "just refresh" is useful.  
(Though, the reason it's particularly useful in the context ov menus is because menus are typically designed to be "reactive", changing at the snap ov a finger – like how websites work.)

> If you want to do it via same-state transitions, you can do it this way with the above code.
> ```rust
> // OLD
> // app.add_systems(OnEnter(AppState::Menu), update_menu);
> // NEW
> app.add_systems(Update, update_menu.run_if(on_any_enter(AppState::Menu));
> ```

I'm guessing you mean "below"(?). — What is `on_any_enter`?  
I looked up ""bevy" "on_any_enter"" on Google and it only showed [the crates.io page for `pyri_state`](https://crates.io/crates/pyri_state/0.1.0).

(I want to make sure this is the same `on_any_enter` that you are talking about.)

**Comment by shanecelis** [comment]
It's [the code I wrote in this comment](https://github.com/bevyengine/bevy/issues/15478#issuecomment-2606339094). It's a `Condition`, which is a `System` that returns a boolean. I'd just copy-and-paste that code and move on to the next thing. If you find it particularly helpful in the long run and think it should be built-in, then you can file a PR for it. I withdraw my initial reservations about it.

**Comment by benfrankel** [comment]
Unfortunately it seems that this particular approach doesn't have enough support (and it's not my preferred approach anyways), so I'm closing this issue. The "parent" issue is still open: https://github.com/bevyengine/bevy/issues/14447.
